### ISSUE TITLE
Onboard to bencher, to start tracking benchmarks over time

### ISSUE THREAD

**Issue by joshka** [ratatui-org/ratatui#1092]
## Problem
Rather than having to run benchmarks ad hoc, It would be nice to see results of our benchmarks over time (particularly in CI) so we can more easily catch any regressions. https://bencher.dev/ seems to be a reasonable product for this with a free hosted tier or open source projects.

## Solution
Onboard to bencher.dev

## Alternatives
Do our own thing and carry the maintenance burden of that.

## Additional context
Anticipating the obvious argument against this that CI is noisy, is covered in their docs: even a noisy signal easily shows when performance has drastically changed (see https://bencher.dev/docs/explanation/continuous-benchmarking/ for more info)

1000 words:
![image](https://github.com/ratatui-org/ratatui/assets/381361/f692d7ad-696e-4658-86de-e283428c82ad)

**Comment by EdJoPaTo** [comment]
Keep in mind that benchmarks are likely biased to the specific problem added in them and might not reflect user usage. It might miss something user code uses that might give a false sense on the benchmark.
This should not stop us from refactoring benchmarks. Having useful benchmarks is way more important than having historic benchmark comparison graphs for longer.

Also, the benchmarks are highly dependent on their environment like target triple. x64 has other benefits compared to different ARM platforms. Apple silicon for example has some instructions explicitly for having easier to run code written for x64 which is why Rosetta works explicitly well on Apple ARM. The Windows ARM comparability layer does not have this which is why it’s way slower. While it’s both ARM it’s hard to compare them because of stuff like that.

As with all benchmarks absolute numbers are only useful on the given target and can not be easily compared. Changes on the other hand can be compared. (Examples: M2 Performance Core with M2 Performance Core is comparable with absolute Numbers. M2 Perf to Efficiency core is not. M2 to random benchmark platform target is not. M2 to Raspberry Pi is not. The change percentages however are comparable at least for the target triple & glibc version.)

So this will only ever give a rough idea and changing a benchmark code will result in a new graph.
This should only ever be a hint at „there might be something of“.

**Comment by joshka** [comment]
> Changes on the other hand can be compared

Yes. This is the entire point of bencher.

**Comment by orhun** [comment]
Looks cool! I'm curious about what do we need to move forward with it? Should we reach out to them or can we just integrate it somehow?

**Comment by joshka** [comment]
> Looks cool! I'm curious about what do we need to move forward with it? Should we reach out to them or can we just integrate it somehow?

Read the docs, sign up, do the tasks required.

**Comment by orhun** [comment]
Doesn't seem free btw

![image](https://github.com/ratatui-org/ratatui/assets/24392180/6c213da0-a8aa-405e-8a25-4d86cb31a1f2)

**Comment by orhun** [comment]
I created a dummy project and uploaded some benchmark results for testing:

```rs
bencher run --project "orhun-s-project" --token $BENCHER_API_TOKEN --adapter rust_criterion "cargo bench"
```

Here is how barchart render benchmark looks like (not very exciting because just a single data point for now):

![image](https://github.com/ratatui-org/ratatui/assets/24392180/45f3d30b-0a60-4ab6-9b68-86e0b657ec64)

[share URL](https://bencher.dev/perf/orhun-s-project?branches=60e1c455-a9e9-44e0-a5ce-20f313ff78c1&testbeds=218e19b0-6c0c-4d0a-a830-03e3b55afb23&benchmarks=de4d2449-98c5-4796-a4c0-6761c2129eee&measures=e40e36c9-d0a3-4b49-9a84-37736f49db11&start_time=1715281032000&end_time=1717873721000&key=true&reports_per_page=4&branches_per_page=8&testbeds_per_page=8&benchmarks_per_page=8&plots_per_page=8&reports_page=1&branches_page=1&testbeds_page=1&benchmarks_page=1&plots_page=1)

I think it's worth experimenting with different options.

**Comment by orhun** [comment]
Okay with 2 data points it is better:

![image](https://github.com/ratatui-org/ratatui/assets/24392180/06511166-bfcf-47b1-aec8-4abc5f947f46)

[url]([https://bencher.dev/perf/orhun-s-project?branches=60e1c455-a9e9-44e0-a5ce-20f313ff78c1&testbeds=218e19b0-6c0c-4d0a-a830-03e3b55afb23&benchmarks=f25e659a-36ed-4512-826f-f2df894350ad&measures=e40e36c9-d0a3-4b49-9a84-37736f49db11&start_time=1715282077000&end_time=1717874809000&upper_boundary=true&key=true&reports_per_page=4&branches_per_page=8&testbeds_per_page=8&benchmarks_per_page=8&plots_per_page=8&reports_page=1&branches_page=1&testbeds_page=1&benchmarks_page=1&plots_page=1](https://bencher.dev/console/projects/orhun-s-project/perf?branches=60e1c455-a9e9-44e0-a5ce-20f313ff78c1&testbeds=218e19b0-6c0c-4d0a-a830-03e3b55afb23&benchmarks=f25e659a-36ed-4512-826f-f2df894350ad&measures=e40e36c9-d0a3-4b49-9a84-37736f49db11&start_time=1715282077000&end_time=1717874809000&upper_boundary=true&key=true&reports_per_page=4&branches_per_page=8&testbeds_per_page=8&benchmarks_per_page=8&plots_per_page=8&reports_page=1&branches_page=1&testbeds_page=1&benchmarks_page=1&plots_page=1))

**Comment by epompeii** [comment]
If there is anything that is a blocker for you all, please let me know!

**Comment by EdJoPaTo** [comment]
Current state is pretty annoying in PR. Long message without actual content seeing diffs and stating a warning. Should be just a link or usable content at first glance.

**Comment by joshka** [comment]
Yeah - I agree to some extend on this - let's see what we can do to iterate on this a bit.

**Comment by epompeii** [comment]
> Current state is pretty annoying in PR. Long message without actual content seeing diffs and stating a warning. Should be just a link or usable content at first glance.

@EdJoPaTo taking a look at the current PR comments, like this one: https://github.com/ratatui-org/ratatui/pull/1252#issuecomment-2264697983, it appears that you all have not set up a [Threshold](https://bencher.dev/docs/explanation/thresholds/) yet. This is what the large warning is about.
Further, without a Threshold setup there won't be any diffs of the Metrics. Once a Threshold is added, the table will look more like the one in this example: https://github.com/bencherdev/bencher#example-pr-comment

I'm also open to feedback on other aspects to improve for the PR comment. Feel free to send them my way!

**Comment by EdJoPaTo** [comment]
> > Current state is pretty annoying in PR. Long message without actual content seeing diffs and stating a warning. Should be just a link or usable content at first glance.
> 
> taking a look at the current PR comments, like this one: [#1252 (comment)](https://github.com/ratatui-org/ratatui/pull/1252#issuecomment-2264697983), it appears that you all have not set up a [Threshold](https://bencher.dev/docs/explanation/thresholds/) yet. This is what the large warning is about.

Yeah, the message is pretty clear about that :innocent: Its more a hint to @orhun to configure or improve it :innocent: 

> Further, without a Threshold setup there won't be any diffs of the Metrics. Once a Threshold is added, the table will look more like the one in this example: https://github.com/bencherdev/bencher#example-pr-comment
> 
> I'm also open to feedback on other aspects to improve for the PR comment. Feel free to send them my way!

Trying to give some feedback, feel free to ask for specifics when something is unclear. Would like to have it as useful as it can be! :)

I think I would prefer an even shorter base message with grouping more things into a `<details>`. Listing the Project / Branch of the PR in the given PR is superfluous, and the whole first table could be hidden / removed from my point of view. Maybe put in into a collapsible together with the footer links to the docs, repo and so on, as they are normally not that interesting but still there to debug / understand / configure / improve things.

Even without the warning it would be double the height of a normal text message in the PR thread while it should most of the time only be relevant when there is an actual regression. Or highlighting that something improves stuff significantly. Otherwise, it should be kind of minimal.

I am mainly interested at the diff (which apparently isn't shown because of the threshold) telling me it's a minimal difference and a link to the public perf page to look at more history / details.

**Comment by joshka** [comment]
Would some default thresholds be reasonable to add as a feature?
Reopening this for discussion.

**Comment by orhun** [comment]
I also agree that the PR comment is a bit long and not super useful with its current state. We can try adding some thresholds and see how it is going to change things.

![image](https://github.com/user-attachments/assets/d623419b-14bb-4ef0-8022-29ebfbf3b489)

Any recommendations about upper/lower limits and the threshold model? @epompeii

I think percentage is the simplest one and it can be good start. We should still figure out some values for it though.

**Comment by EdJoPaTo** [comment]
`criterion` seems to use 5% below or above the last result as relevant. So maybe just use that?

**Comment by epompeii** [comment]
The [Percentage Threshold](https://bencher.dev/docs/explanation/thresholds/#percentage-thresholds) is definitely one of the simplest, however it is likely to lead to more false positives.

> `criterion` seems to use 5% below or above the last result as relevant. So maybe just use that?

In order to accomplish this, you would need to set the max sample size to `2` and the upper boundary to `0.05`. However, I would not recommend this approach. It will be _very_ noisy.

> Would some default thresholds be reasonable to add as a feature?

The default that is setup for every project's initial dimensions uses the [Student's t-test Threshold](https://bencher.dev/docs/explanation/thresholds/#t-test-thresholds). The Threshold Model is configured with a max sample size of `255` and an upper boundary of `0.98`. I would recommend using this as your starting point.

A default Threshold is not created for every possible permutation of every dimension (Branch x Testbed x Measure) due to the combinatorial explosion that would entail.

> Trying to give some feedback, feel free to ask for specifics when something is unclear. Would like to have it as useful as it can be! :)

Thank you for the specific feedback!
I have created a tracking issue: https://github.com/bencherdev/bencher/issues/473

**Comment by joshka** [comment]
Added more feedback on metrics display at https://github.com/bencherdev/bencher/issues/474

**Comment by orhun** [comment]
[CodSpeed](https://codspeed.io) had a booth at EuroRust - they seem like an interesting project with a similar flavor to Bencher 👀

**Comment by epompeii** [comment]
> they seem like an interesting project with a similar flavor to Bencher 

Definitely similar motivations!

If you're interested in being able to track instruction counts (which is what CodSpeed is doing in CI) you may also want to check out:
- `iai`:
  - repo: https://github.com/bheisler/iai
  - adapter: https://bencher.dev/docs/explanation/adapters/#-rust-iai
Or
- iai-callgrind:
  - repo: https://github.com/iai-callgrind/iai-callgrind
  - adapter: https://bencher.dev/docs/explanation/adapters/#-rust-iai-callgrind

I obviously have a bit of bias here. However, I just wanted to make sure to point out that only tracking and comparing instruction count based benchmarks can lead to a false sense of confidence. CodSpeed uses a proprietary algorithm for weighting these instruction counts, but it is still only a proxy for actual speed, wall clock time. Further, this "estimation of speed" does not include any time spent in syscalls, which can lead to _extreme_ differences in real world performance: https://docs.codspeed.io/features/understanding-the-metrics/#system-calls

I think CodSpeed is great for micro-benchmarking pure algorithms. Otherwise, I think it's worth tracking both wall clock and instruction count based benchmarks.

Does that make sense?

It's great to see they're out there promoting continuous benchmarking!

**Comment by orhun** [comment]
Thanks for the insight! Yes, that pretty much makes sense.

**Comment by joshka** [comment]
Closing this out for now to reduce the backlog to just things that we're actively working on.

The main observation we formed was that using this was a bit too noisy for our use cases when there are many many benchmarks. We might revisit things in the future.
